storage/internal/benchmarks/directory_benchmark.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"math/rand"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/api/iterator"
+)
+
+// directory benchmark is a benchmark that uploads a randomly generated
+// directory once and then downloads it once. The objects in the directory may
+// be uploaded/downloaded using parallelization.
+//
+// SET UP
+//   - Initialize structs to populate with the benchmark results.
+//   - Select a random API to use to upload/download the object, unless it is
+//     set in the command-line,
+//   - Select a random size for the objects that will be uploaded; this size will
+//     be between two values configured in the command-line.
+//   - Create a directory with 1 to 1000 files of that size and subdirectories.
+//   - Select, for the upload and the download separately, the following parameters:
+//   - the application buffer size set in the command-line
+//   - the chunksize (only for uploads) set in the command-line,
+//   - if the client library will perform CRC32C and/or MD5 hashes on the data.
+//
+// BENCHMARK
+//   - Grab a storage client from the pool.
+//   - Take a snapshot of the current memory stats.
+//   - Upload the entire directory, capturing the time taken.
+//   - Take another snapshot of memory stats.
```

Comment:
It would be more useful to be able to capture memory usage as the operation runs. Can this perhaps run in a separate Go routine on a timer and capture stats at fixed intervals?

Reply:
I agree, really I haven't put much focus on the capturing or analysis of memory stats. Do you have a proposal for how memory stats should be output? Maybe we could add more metrics (would need a bit of investigation to make sure it can be done and captured by the prober framework; although for current workloads running on the framework I don't know how useful the data would actually be)?

Reply:
I don't have a fully fleshed out proposal for a metric, but I think what would be useful to capture is average & max memory usage while the workload is running, based on snapshots taken at (say) a 60s interval. But, I think it's beyond the scope of this PR -- could you file a separate issue to add this?